CMakeLists.txt

```diff
@@ -14,24 +14,43 @@ find_package(Python3 COMPONENTS Interpreter REQUIRED)
 
 #We have Python, now let's check to see if we have setuptools
 find_python_module(setuptools REQUIRED)
+find_python_module(venv REQUIRED)
 
-if (PY_SETUPTOOLS)
+set(SETUP_PY_IN "${CMAKE_CURRENT_SOURCE_DIR}/setup.py") #setup.py has logic to automatically switch between CMake variables and normal operation
+set(SETUP_PY    "${CMAKE_CURRENT_BINARY_DIR}/setup.py")
+set(DEPS        "${CMAKE_CURRENT_SOURCE_DIR}/cmakedoc/__init__.py")
+set(OUTPUT      "${CMAKE_CURRENT_BINARY_DIR}/build/timestamp")
 
-    set(SETUP_PY_IN "${CMAKE_CURRENT_SOURCE_DIR}/setup.py") #setup.py has logic to automatically switch between CMake variables and normal operation
-    set(SETUP_PY    "${CMAKE_CURRENT_BINARY_DIR}/setup.py")
-    set(DEPS        "${CMAKE_CURRENT_SOURCE_DIR}/cmakedoc/__init__.py")
-    set(OUTPUT      "${CMAKE_CURRENT_BINARY_DIR}/build/timestamp")
 
-    configure_file(${SETUP_PY_IN} ${SETUP_PY})
+configure_file(${SETUP_PY_IN} ${SETUP_PY})
 
-    add_custom_command(OUTPUT ${OUTPUT}
-                       COMMAND ${Python3_EXECUTABLE} ${SETUP_PY} build
-                       COMMAND ${CMAKE_COMMAND} -E touch ${OUTPUT}
-                       DEPENDS ${DEPS})
+set(VENV "${CMAKE_CURRENT_BINARY_DIR}/virtual-env")
 
-    add_custom_target(target ALL DEPENDS ${OUTPUT})
+execute_process(COMMAND "${Python3_EXECUTABLE}" "-m" 
+                        "venv" "${VENV}"
+                        RESULT_VARIABLE _venv_status 
+                        ERROR_QUIET OUTPUT_STRIP_TRAILING_WHITESPACE)
 
-    install(CODE "execute_process(COMMAND ${PYTHON} ${SETUP_PY} install)")
-else()
-    message(FATAL_ERROR "setuptools Python3 package not installed")
+if (NOT _venv_status EQUAL "0")
+	message(FATAL_ERROR "Virtual environment could not be created, venv returned ${_venv_status}")
 endif()
+
+set(VENV_PYTHON_EXECUTABLE "${VENV}/bin/python3")
+set(VENV_PIP "${VENV}/bin/pip3")
+
+execute_process(COMMAND "${VENV_PIP}" "install" "pyinstaller" "wheel") #Install our build tool to our build environment
```

Comment:
does it make sense to just write a `requirements.txt` file and have PIP install that way, i.e., `pip install -r requirements.txt`?

Reply:
For now, that would work just fine. The problem with requirements.txt is that it can contain metadata about packages, such as index URLs or sections for package extras. This would mean that to be safe we would need to parse the file in setup.py, rather than just splitting on lines, and parsing requirements.txt is different for different versions of pip and newer than 6.0 requires a hack.

Reply:
We would be the ones writing the file so we can make sure we write the file without metadata. Perhaps I'm thinking to simply, but I assumed it would be as simple as:
```
pyinstaller==x.y
wheel==x.y
antlr==x.y
...
```

Reply:
Yes that's the general form, the additional metadata is used by pip in uncommon use cases, but as long as we don't use those esoteric features we should be fine. I'll split dependencies out to a requirements.txt tonight.

Reply:
Should be fixed now